amaranth/build/plat.py

```diff
@@ -315,6 +315,45 @@ def toolchain_prepare(self, fragment, name, **kwargs):
 
         rtlil_text, self._name_map = rtlil.convert_fragment(fragment, name=name)
 
+        # Retrieve an override specified in either the environment or as a kwarg.
+        # expected_type parameter is used to assert the type of kwargs, passing `None` will disable
+        # type checking.
+        def _extract_override(var, *, expected_type):
+            deprecated_var_env = "NMIGEN_{}".format(var)
+            var_env = "AMARANTH_{}".format(var)
+            if deprecated_var_env in os.environ or var_env in os.environ:
+                # On Windows, there is no way to define an "empty but set" variable; it is tempting
+                # to use a quoted empty string, but it doesn't do what one would expect. Recognize
+                # this as a useful pattern anyway, and treat `set VAR=""` on Windows the same way
+                # `export VAR=` is treated on Linux.
+                if var_env in os.environ:
+                    var_env_value = os.environ[var_env]
+                elif deprecated_var_env in os.environ:
+                    var_env_value = os.environ[deprecated_var_env]
+                return re.sub(r'^\"\"$', "", var_env_value)
+            elif var in kwargs:
+                if expected_type is None or isinstance(kwargs[var], expected_type):
+                    return kwargs[var]
+                else:
+                    raise TypeError("Invalid keyword argument '{}' of type '{}' when type '{}' was expected."
+                                    .format(var, type(kwargs[var]).__name__, expected_type.__name__))
```

Comment:
Our type errors are usually more like:

    raise TypeError("Override '{}' must be a {}, not {!r}".format(var, expected_type.__name__, kwargs[var])

Aside from the phrasing differences this shows you what actual object you got in there, not just its type, which is convenient.

A minor stylistic issue, but if this was converted to an early raise, then every branch would lexically end with a `return`.

Reply:
Agreed. I didn't search hard enough when looking for example error messages to follow. I also agree with the stylistic change.